src/main/java/picard/nio/PicardHtsPath.java

```diff
@@ -117,4 +121,45 @@ public static List<Path> toPaths(final Collection<PicardHtsPath> picardHtsPaths)
         Objects.requireNonNull(picardHtsPaths);
         return picardHtsPaths.stream().map(PicardHtsPath::toPath).collect(Collectors.toList());
     }
+
+    /**
+     * Takes an IOPath and returns a new PicardHtsPath object that keeps the same basename as the original but has
+     * a new extension. If the input path is a relative local path, it is converted to an absolute path via
+     * PicardHtsPath::toPath (we might change this behavior and avoid converting to an absolute path in the future,
+     * by for instance working directly with a string).
+     *
+     * Examples:
+     *     - (test_na12878.bam, .bai) -> test_na12878.bai (append = false)
+     *     - (test_na12878.bam, .md5) -> test_na12878.bam.md5 (append = true)
+     *
+     * @param path The original path
+     * @param append If set to true, append the new extension to the original basename. If false, replace the original extension
+     *               with the new extension. If append = false and the original name has no extension, an exception will be thrown.
+     * @param newExtension A new file extension. Must include the leading dot e.g. ".txt", ".bam"
+     * @return A new PicardHtsPath object with the new extension
+     */
+    public static PicardHtsPath replaceExtension(final IOPath path, final String newExtension, final boolean append){
+        ValidationUtils.validateArg(newExtension.startsWith("."), "newExtension must start with a dot '.'");
+
+        if (append){
+            return new PicardHtsPath(path.getURIString() + newExtension);
+        } else {
+            final Optional<String> oldExtension = path.getExtension();
+
+            if (oldExtension.isEmpty()){
+                throw new PicardException("The original path must have an extension when append = false: " + path.getURIString());
+            }
+
+            final String oldFileName = path.toPath().getFileName().toString();
+            final String newFileName = oldFileName.replaceAll(oldExtension.get() + "$", newExtension);
+            return PicardHtsPath.fromPath(path.toPath().resolveSibling(newFileName));
+        }
```

Comment:
The `append==true` code path is still doing string manipulation. It would be preferable to handle both code paths the way you're handling the `append==false` case now, using `resolveSibling`. I can't test this since I don't have your modified htsjdk, but it would look something like this:
```suggestion
        final String oldFileName = path.toPath().getFileName().toString();
 
        String newFileName;
        if (append){
            newFileName = oldFileName + newExtension;
        } else {
            final Optional<String> oldExtension = path.getExtension();
            if (oldExtension.isEmpty()){
                throw new PicardException("The original path must have an extension when append = false: " + path.getURIString());
              newFileName = oldFileName.replaceAll(oldExtension.get() + "$", newExtension);
            }
            return PicardHtsPath.fromPath(path.toPath().resolveSibling(newFileName));
        }
```
The litmus test would be a URI test case that looks like `http://some/file.txt?someParam=someValue`, which I suspect the current code handles correctly for `append==false` but incorrectly for `append==true`. It would be great to add those test cases.

Reply:
I think string manipulation cannot be avoided here (your proposed code does string manipulation too), and it seems to me that the current code is more natural and easier to understand. That said, I also see your point that the current code may not work for input URI like `http://some/file.txt?someParam=someValue`. But my questions then are:

1) should we support such a file path? If we support http:// paths at all, isn't it reasonable to ask the user to clean up the path before feeding it to Picard?
2) given that we should handle `http://some/file.txt?someParam=someValue` correctly, what should be the output path, with append == true and newExtension = ".log"? `http://some/file.txt.log`? `http://some/file.txt.log?someParam=someValue`?

Reply:
Sounds good, will try to support query params. The only issue is that when I use the http:// URI in the text, I get "ProviderNotFoundException, Provider 'http' not found", due to the fact that in `Path:of`, `FileSystemProvider.installedProviders()` returns a list that does not include the scheme "http". Will investigate.

Reply:
So I looked into this a bit further and found the following:

- "http://" filesystem is not currently supported in Picard (as mentioned during the group meeting). So adding test cases for the http:// URI's, with or without query params, cannot be done. htsjdk has test cases with http:// because those test cases are for the URI objects, which can parse the rawInputString independent of the filesystem. We run into trouble only when we create the filesystem-specific implementation of the Path interface (e.g. UnixPath, CloudStoragePath)
- Test cases with query params attached to a local or gs:// URI fails also, with explicit errors about query params not being supported thrown by the corresponding Path objects. (I can see why local file path does not support query params, but I thought gs:// should, in which case this behavior might be a bug…)

The upshot is, in the current state of Picard, we cannot support query params and so we can't write functioning tests with query params. The first step to supporting query params is to support the `http://` URI, which I'd be happy to do as my next project.